polkadot/transaction-pool/src/lib.rs

```diff
@@ -44,19 +47,66 @@ pub fn truncate_id(id: &AccountId) -> TruncatedAccountId {
 	TruncatedAccountId::from_slice(&id[..20])
 }
 
+/// Useful functions for working with Polkadot blocks.
+pub struct PolkadotBlock (Block);
+
+impl PolkadotBlock {
+	/// Create a new block, checking high-level well-formedness.
+	pub fn from(unchecked: Block) -> ::std::result::Result<Self, Block> {
+		if unchecked.extrinsics.len() < 1 {
+			return Err(unchecked);
+		}
+		if unchecked.extrinsics[0].is_signed() {
+			return Err(unchecked);
+		}
+		if let Call::Timestamp(TimestampCall::set(_)) = unchecked.extrinsics[0].extrinsic.function {} else {
+			return Err(unchecked);
+		}
+		Ok(Self::force_from(unchecked))
+	}
+
+	/// Create a new block, skipping any high-level well-formedness checks. WARNING: This could
+	/// result in internal functions panicking if the block is, in fact, not well-formed.
+	pub fn force_from(known_good: Block) -> Self {
```

Comment:
the main problem with this is that the errors occur in such a way that it's difficult to track down the point at which the programmer error was actually made. 

We're changing types like `HeaderView` which are supposed to take only trusted data in Parity-Ethereum to solve a similar problem by having the function take the file name and line number and having the view carry those around to incorporate in panic messages, so we get the cause instead of just the symptom. We also provide a macro which invokes this constructor using the `file!()` and `line!()` macros internally.

Reply:
happy to place an issue and bring it in in a later PR.